[UnifiedPDF] [iOS] Navigate when tapping on links in PDF documents

whsieh opened the pull request:
#### 9e046abb49e52a1190018fc9a6c167e70c73f4eb
<pre>
[UnifiedPDF] [iOS] Navigate when tapping on links in PDF documents
<a href="https://bugs.webkit.org/show_bug.cgi?id=284719">https://bugs.webkit.org/show_bug.cgi?id=284719</a>
<a href="https://rdar.apple.com/141519205">rdar://141519205</a>

Reviewed by Richard Robinson.

Add support for tapping links in PDF documents when unified PDF is enabled (along with some
additional items). See below for more details.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::highlightRectForTapAtPoint const):
(WebKit::PDFPluginBase::handleSyntheticClick):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::navigateToURL):

Add an optional `PlatformMouseEvent` parameter, for use when producing the `MouseEvent` used to
trigger the navigation. This allows a new tab to be opened, when holding down ⌘ while tapping a link
in a PDF.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::followLinkAnnotation):
(WebKit::UnifiedPDFPlugin::highlightRectForTapAtPoint const):

Add a new PDF plugin hook to return the highlight rect at the given point in root view coordinates.

(WebKit::UnifiedPDFPlugin::handleSyntheticClick):

Add a method to handle synthetic clicks (taps) on the PDF plugin. This will do one of three things:

- Extend the current selection if the shift key is held
- Follow a link under the given point
- Clear the active selection

(WebKit::UnifiedPDFPlugin::platformPopulateEditorStateIfNeeded const):
(WebKit::UnifiedPDFPlugin::pageToRootView const):
(WebKit::UnifiedPDFPlugin::rootViewToPage const):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::highlightRectForTapAtPoint const):
(WebKit::PluginView::handleSyntheticClick):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::completeSyntheticClick):

Change the call to `clearSelection()` into `handleSyntheticClick()`, and pass in the whole platform
event (see above).

(WebKit::WebPage::sendTapHighlightForNodeIfNecessary):

Defer to the plugin if needed, when retrieving tap highlight geometry.

(WebKit::WebPage::handleTwoFingerTapAtPoint):
(WebKit::WebPage::potentialTapAtPosition):
(WebKit::WebPage::tapHighlightAtPosition):

Canonical link: <a href="https://commits.webkit.org/287883@main">https://commits.webkit.org/287883@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8126f97040b082aae86115bbcb31428a1a253374

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81159 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/684 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35102 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85688 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32145 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/702 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8494 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63362 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21129 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84228 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/451 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73863 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43660 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/348 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30603 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71871 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87123 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8389 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5950 "Found 8 new test failures: compositing/tiling/tiled-reflection-inwindow.html fast/forms/color/color-input-reset-crash.html fast/forms/color/color-type-change-on-input-crash.html fast/forms/color/input-color-readonly.html http/tests/navigation/page-cache-pending-ping-load-same-origin.html imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/show-picker-user-gesture.html media/non-existent-video-playback-interrupted.html tiled-drawing/tiled-drawing-scroll-position-page-cache-restoration.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71667 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8568 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69698 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70902 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17664 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14948 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13869 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8350 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13874 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8187 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11707 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9995 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -1144,8 +1144,10 @@
-        coreEvent = MouseEvent::create(eventNames().clickEvent, &frame->windowProxy(), platform(*m_lastMouseEvent), { }, { }, 0, 0);
+    if (event || m_lastMouseEvent) {
+        auto platformEvent = event ? WTFMove(*event) : platform(*m_lastMouseEvent);
+        coreEvent = MouseEvent::create(eventNames().clickEvent, &frame->windowProxy(), platformEvent, { }, { }, 0, 0);
```
(I know this isn't new to this patch but it just reminded me, this should probably be changed to be a PointerEvent at some point)

whsieh replied:
Makes sense! Perhaps separately, as a followup patch..

whsieh committed:
[UnifiedPDF] [iOS] Navigate when tapping on links in PDF documents
https://bugs.webkit.org/show_bug.cgi?id=284719
rdar://141519205

Reviewed by Richard Robinson.

Add support for tapping links in PDF documents when unified PDF is enabled (along with some
additional items). See below for more details.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::highlightRectForTapAtPoint const):
(WebKit::PDFPluginBase::handleSyntheticClick):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::navigateToURL):

Add an optional `PlatformMouseEvent` parameter, for use when producing the `MouseEvent` used to
trigger the navigation. This allows a new tab to be opened, when holding down ⌘ while tapping a link
in a PDF.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::followLinkAnnotation):
(WebKit::UnifiedPDFPlugin::highlightRectForTapAtPoint const):

Add a new PDF plugin hook to return the highlight rect at the given point in root view coordinates.

(WebKit::UnifiedPDFPlugin::handleSyntheticClick):

Add a method to handle synthetic clicks (taps) on the PDF plugin. This will do one of three things:

- Extend the current selection if the shift key is held
- Follow a link under the given point
- Clear the active selection

(WebKit::UnifiedPDFPlugin::platformPopulateEditorStateIfNeeded const):
(WebKit::UnifiedPDFPlugin::pageToRootView const):
(WebKit::UnifiedPDFPlugin::rootViewToPage const):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::highlightRectForTapAtPoint const):
(WebKit::PluginView::handleSyntheticClick):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::completeSyntheticClick):

Change the call to `clearSelection()` into `handleSyntheticClick()`, and pass in the whole platform
event (see above).

(WebKit::WebPage::sendTapHighlightForNodeIfNecessary):

Defer to the plugin if needed, when retrieving tap highlight geometry.

(WebKit::WebPage::handleTwoFingerTapAtPoint):
(WebKit::WebPage::potentialTapAtPosition):
(WebKit::WebPage::tapHighlightAtPosition):

Canonical link: https://commits.webkit.org/287883@main

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3889,6 +3889,54 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+#if HAVE(PDFDOCUMENT_SELECTION_WITH_GRANULARITY)
+    auto pointInRootView = event.position();
+    RetainPtr selection = m_currentSelection;
+    if (selection && event.shiftKey()) {
```
Should we allow fallthrough here even with an active selection? Currently we do navigate when shift-clicking on a link annotation while a selection is active

whsieh replied:
Makes sense — I can make that adjustment in my next PR...

(Actually, I'll move the link navigation code above, to give it priority)
